docs/architecture/psa-keystore-design.md

```diff
@@ -0,0 +1,204 @@
+
+#### Dynamic key slot deallocation
+
+When destroying a volatile key, `psa_wipe_key_slot()` calls `psa_free_key_slot()`. This function adds the newly freed slot to the head of the free list.
+
+##### Future improvement: slice deallocation
+
+As of Mbed TLS 3.6.1, `psa_free_key_slot()` does not deallocate slices. Thus the memory consumption for slots never decreases (except when the PSA crypto subsystem is deinitialized). Freeing key slices intelligently would be a desirable improvement.
+
+We should not free a key slice as soon as it becomes empty, because that would cause large allocations and deallocations if there are slices full of long-lived keys, and then one slice keeps being allocate and deallocated for the occasional short-lived keys. Rather, there should be some hysteresis, e.g. only deallocate a slice if there are at least T free slots in the previous slice.
+
+Note that currently, the slice array contains one sequence of allocated slices followed by one sequence of unallocated slices. Mixing allocated and unallocated slices may make some parts of the code a little more complex, and should be tested thoroughly.
+
+### Persistent key cache
+
+Persistent keys and built-in keys need to be loaded into the in-memory key store each time they are accessed:
+
+* while creating them;
+* to access their metadata;
+* to start performing an operation with the key;
+* when destroying the key.
+
+To avoid frequent storage access, we cache persistent keys in memory. This cache also applies to built-in keys.
+
+With the [static key store](#static-key-store), a non-empty slot can contain either a volatile key or a cache entry for a persistent or built-in key. With the [dynamic key store](#dynamic-key-store), volatile keys and cached keys are placed in separate [slices](#key-slices-in-the-dynamic-key-store).
+
+The persistent key cache is a fixed-size array of `MBEDTLS_PSA_KEY_SLOT_COUNT` slots. This array is shared with volatile keys in the static key store, and separate with the dynamic key store.
+
+#### Accessing a persistent key
+
+`psa_get_and_lock_key_slot()` automatically loads persistent and built-in keys if the specified key identifier is in the corresponding range. To that effect, it traverses the key cache to see if a key with the given identifier is already loaded. If not, it loads the key. This cache walk takes time that is proportional to the cache size.
+
+#### Slot readers
+
+At any given time, a slot can have registered readers. The `registered_readers` fields contains the number of readers. The following are readers:
+
+* A thread that is currently reading metadata or data from the slot (see [“Concurrency”](#concurrency)).
+* A persistent key that has been opened with `psa_open_key()`, until the corresponding handle is closed with `psa_close_key()`.
```

Comment:
Hi Gilles! I saw this PR on slack and wanted to give it a quick scan through. I spotted the following:

This second bullet point is not true. `psa_open_key` effectively just calls `psa_get_and_lock_key_slot` followed by `psa_unregister_read`. On return the caller is not counted as a "registered reader". The handle to the key which they are given is simply the key identifier (which they already passed into the call), this can be plugged into operations just like any other key id. The matching call to `psa_close_key` will just try to remove the key from the slot which it currently occupies.

Before returning, all non internal functions (including multi-part operation calls) unregister from every slot that they have registered to. This prevents the possibility of infinite waits when running multithreaded calls, and other potentially unforseen headaches.